primitives/src/unique_vec.rs

```diff
@@ -0,0 +1,363 @@
+    pub fn push(&mut self, value: T) -> bool {
+        if self.contains(&value) {
+            false
+        } else {
+            self.0.push(value);
+            true
+        }
+    }
+
+    /// Removes the element at the given `index` and returns it.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the `index` is out of bounds.
+    pub fn remove(&mut self, index: usize) -> T {
+        self.0.remove(index)
+    }
+
+    /// Clears the [`UniqueVec`], removing all values.
+    pub fn clear(&mut self) {
+        self.0.clear()
+    }
+}
+
+impl<T: PartialEq> FromIterator<T> for UniqueVec<T> {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        let mut unique_vec = Self::new();
+        unique_vec.extend(iter);
+        unique_vec
+    }
+}
+
+impl<T: PartialEq> Extend<T> for UniqueVec<T> {
+    fn extend<I: IntoIterator<Item = T>>(&mut self, iter: I) {
+        for value in iter {
+            self.push(value);
+        }
+    }
+}
+
+impl<T: PartialEq> From<UniqueVec<T>> for Vec<T> {
+    fn from(value: UniqueVec<T>) -> Self {
+        value.0
+    }
+}
+
+impl<T: PartialEq> AsRef<[T]> for UniqueVec<T> {
+    fn as_ref(&self) -> &[T] {
+        self.0.as_ref()
+    }
+}
+
+impl<T: PartialEq> Borrow<[T]> for UniqueVec<T> {
+    fn borrow(&self) -> &[T] {
+        self.0.borrow()
+    }
+}
+
+impl<T: PartialEq> Deref for UniqueVec<T> {
+    type Target = [T];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T: PartialEq> IntoIterator for UniqueVec<T> {
+    type Item = T;
+    type IntoIter = <Vec<T> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<'ve, T: PartialEq> IntoIterator for &'ve UniqueVec<T> {
+    type Item = &'ve T;
+    type IntoIter = <&'ve Vec<T> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
+}
+
+impl<'de, T: PartialEq + Deserialize<'de>> UniqueVec<T> {
+    /// Deserialize [`UniqueVec`], failing on first duplicate.
+    ///
+    /// Default implementation of [`Deserialize`] for [`UniqueVec`] ignores duplicates.
+    ///
+    /// # Errors
+    ///
+    /// - If deserialization of `T` fails.
+    /// - If there are duplicates in the sequence.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use serde::{Deserialize, de::Error as _};
+    /// use iroha_primitives::unique_vec::UniqueVec;
+    ///
+    /// #[derive(Debug, PartialEq, Deserialize)]
+    /// pub struct Config {
+    ///     #[serde(deserialize_with = "UniqueVec::deserialize_failing_on_duplicates")]
+    ///     numbers: UniqueVec<u32>,
+    /// }
+    ///
+    /// let err = serde_json::from_str::<Config>(r#"{"numbers": [1, 2, 3, 2, 4, 5]}"#).unwrap_err();
+    /// assert_eq!(
+    ///     err.to_string(),
+    ///     "Duplicated value at line 1 column 25",
+    /// );
+    /// ```
+    pub fn deserialize_failing_on_duplicates<D>(deserializer: D) -> Result<Self, D::Error>
```

Comment:
IMHO these specialized deserialization functions should be defined where they are used. In this case it's the config

Reply:
I don't see the reason to make it less flexible if we already have more generic implementation...

@Erigara ?

Reply:
> Would orphan rule allow this? (we anyway can workaround this by using plain function)

Yes, as you said, we can use plain function, and that's how `#[serde(deserialize_with = "foo")]` works.

> Do we expect that this funtion can be reused in other places except config, if yes i would leave it here.

I don't know, maybe someday... But in that day someone who will write this code may not know that such function is already implemented in config and will re-implement it

Reply:
I don't mind leave it here.